### PR TITLE
Fix attributes collecting

### PR DESCRIPTION
**Pull request by xepozz**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues  | https://github.com/yiisoft/validator/issues/243

- [X] Each rules can be used as an attribute
- [X] Even multiple times
- [X] `HasOne` and `HasMany` were replaced with `Embedded`

I've added TODO about collecting attributes in lazy mode. It may be useful because of using `Reflection` to get rules collection.

### CODE REVIEW

**Comment by xepozz** [comment]
BTW, could someone fix psalm issue? :)

File: src/Attribute/Embedded.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    public function getRuleSet(): array
+    {
+        $classMeta = new ReflectionClass($this->referenceClassName);
+
+        return $this->collectAttributes($classMeta);
+    }
+
+    // TODO: use Generator to collect attributes
```
Comment (samdark) [inline src/Attribute/Embedded.php:41]: ?

File: src/DataSet/AttributeDataSet.php [hunk 2]
```diff
@@ -36,82 +31,19 @@ public function getRules(): iterable
     {
         $classMeta = new ReflectionClass($this->baseAnnotatedObject);
 
-        return $this->handleAttributes($classMeta);
+        return $this->collectAttributes($classMeta);
     }
 
-    private function handleAttributes(ReflectionClass $classMeta): array
+    // TODO: use Generator to collect attributes
```
Comment (samdark) [inline src/DataSet/AttributeDataSet.php:37]: ?

**Comment by andrew-demb** [comment]
@xepozz here it is

```diff
diff --git a/src/Attribute/Embedded.php b/src/Attribute/Embedded.php
index eb80e82..a568ad8 100644
--- a/src/Attribute/Embedded.php
+++ b/src/Attribute/Embedded.php
@@ -18,14 +18,14 @@ use Yiisoft\Validator\ValidationContext;
 #[Attribute(Attribute::TARGET_PROPERTY | Attribute::IS_REPEATABLE)]
 final class Embedded extends GroupRule
 {
+    /**
+     * @psalm-param Closure(mixed, ValidationContext):bool|null $when
+     */
     public function __construct(
         private string $referenceClassName,
         string $message = 'This value is not a valid.',
         bool $skipOnEmpty = false,
         bool $skipOnError = false,
-        /**
-         * @psalm-param Closure(mixed, ValidationContext):bool|null
-         */
         ?Closure $when = null,
     ) {
         parent::__construct($message, $skipOnEmpty, $skipOnError, $when);
diff --git a/src/Rule/GroupRule.php b/src/Rule/GroupRule.php
index 46dfccb..31f4ea6 100644
--- a/src/Rule/GroupRule.php
+++ b/src/Rule/GroupRule.php
@@ -27,7 +27,7 @@ abstract class GroupRule implements ParametrizedRuleInterface, BeforeValidationI
         private bool $skipOnEmpty = false,
         private bool $skipOnError = false,
         /**
-         * @psalm-param Closure(mixed, ValidationContext):bool|null
+         * @var Closure(mixed, ValidationContext):bool|null
          */
         private ?Closure $when = null,
     ) {

```

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Looks way better than current HasOne and HasMany 👍

**Comment by samdark** [comment]
@xepozz seems there are issues not only with Psalm: https://github.com/yiisoft/validator/runs/7368352855?check_suite_focus=true

**Comment by samdark** [comment]
This one includes https://github.com/yiisoft/validator/pull/244, right?

**Comment by samdark** [comment]
Also it seems it will solve https://github.com/yiisoft/validator/pull/193, right?

**Comment by xepozz** [comment]
Not sure about the last one, but the first has to be removed after the merge

File: src/Attribute/Embedded.php (hunk 1 above)
Comment (xepozz) [reply src/Attribute/Embedded.php:41]: I've mentioned it in the PR description

**Comment by xepozz** [comment]
> @xepozz seems there are issues not only with Psalm: https://github.com/yiisoft/validator/runs/7368352855?check_suite_focus=true

Seems nested attributes don't work below 8.1 php version
https://3v4l.org/Dj0lN

File: src/DataSet/AttributeDataSet.php [hunk 3]
```diff
@@ -36,82 +31,19 @@ public function getRules(): iterable
...
-                        [$nestedRule],
-                        ...(($property->getAttributes(Each::class)[0] ?? null)?->getArguments() ?? [])
-                    );
-                }
-            }
-
-            $eachRuleFound = false;
-            $eachRules = [];
-            $attributes = $property->getAttributes();
+            $attributes = $property->getAttributes(RuleInterface::class, ReflectionAttribute::IS_INSTANCEOF);
```
Comment (rustamwin) [inline src/DataSet/AttributeDataSet.php:42]: Now we collect static properties as well?

File: src/Attribute/Embedded.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    private function collectAttributes(ReflectionClass $classMeta): array
+    {
+        $rules = [];
+        foreach ($classMeta->getProperties() as $property) {
+            $attributes = $property->getAttributes(RuleInterface::class, ReflectionAttribute::IS_INSTANCEOF);
+            foreach ($attributes as $attribute) {
+                /** @psalm-suppress UndefinedMethod */
```
Comment (rustamwin) [inline src/Attribute/Embedded.php:48]: It's better to avoid psalm-suppression or comment the reason.

File: src/DataSet/AttributeDataSet.php (hunk 3 above)
Comment (xepozz) [reply src/DataSet/AttributeDataSet.php:42]: yep

**Comment by samdark** [comment]
1. Need to mention PHP 8.0 limitations in readme.
2. Need it to be usable as code, not only as an attribute.

**Comment by samdark** [comment]
PHP 8.1 specific tests should be separated.

File: src/Attribute/Embedded.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    public function getRuleSet(): array
+    {
+        $classMeta = new ReflectionClass($this->referenceClassName);
+
+        return $this->collectAttributes($classMeta);
+    }
+
+    // TODO: use Generator to collect attributes
```
Comment (samdark) [inline src/Attribute/Embedded.php:41]: Why if we return ruleset as array? We would not gain anything, I think.

**Review by samdark — APPROVED** [review]

File: src/Attribute/Embedded.php (hunk 1 above)
Comment (xepozz) [reply src/Attribute/Embedded.php:41]: https://github.com/yiisoft/validator/issues/261

**Pull request merged**